### PR TITLE
Elastic training implementation based on easyscale framework for pytorch job.

### PR DESCRIPTION
**Pull request by SimonCqk**
### Ⅰ. Describe what this PR does

1. implement elastic training protocal(easyscale) on pytorch.
2. introduce a new role: AIMaster.

### CODE REVIEW

File: pkg/util/concurrent/concurrent.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package concurrent
+
+import "sync"
+
+func NewSemaphere(tickets int) *Semaphere {
```
Comment (jian-he) [inline pkg/util/concurrent/concurrent.go:5]: should be semaphore ?

File: controllers/pytorch/pytorchjob_controller.go [hunk 2]
```diff
@@ -230,7 +252,18 @@ func (r *PytorchJobReconciler) SetClusterSpec(ctx context.Context, job interface
 		rank++
 	}
 
-	totalReplicas := int(k8sutil.GetTotalReplicas(pytorchJob.Spec.PyTorchReplicaSpecs))
+	totalReplicas := int(k8sutil.GetTotalExcludedReplicas(pytorchJob.Spec.PyTorchReplicaSpecs, v1.JobReplicaTypeAIMaster))
+	enableElasticScaling := pytorchJob.Annotations[v1.AnnotationEnableElasticTraining] == "true"
+	if enableElasticScaling && !masterRole && rtype != "aimaster" {
+		AddImageWarmupForWorker(podTemplate, r.GetDefaultContainerName())
```
Comment (jian-he) [inline controllers/pytorch/pytorchjob_controller.go:258]: Is this trying to warmup the worker while waiting for master to be ready, but kubelet will run initContainer sequentially.

File: pkg/util/concurrent/concurrent.go (hunk 1 above)
Comment (SimonCqk) [reply pkg/util/concurrent/concurrent.go]: done

File: controllers/pytorch/pytorchjob_controller.go (hunk 2 above)
Comment (SimonCqk) [reply controllers/pytorch/pytorchjob_controller.go:258]: it's true that kubelet will run initContainers sequentially first, warmup container does nothing but pull image of pytorch container, then the next initContainer in order will wait for master to be ready.

**Comment by jian-he** [comment]
lgtm

**Pull request merged**